yarl/_url.py

```diff
@@ -1124,28 +1124,25 @@ def join(self, url):
             raise TypeError("url should be URL")
         other: URL = url
         scheme = other.scheme or self.scheme
-        parts = {
-            k[4:]: getattr(self, k) or ""
-            for k in ("raw_authority", "raw_path", "raw_query_string", "raw_fragment")
-        }
-        parts["scheme"] = scheme
 
         if scheme != self.scheme or scheme not in uses_relative:
-            return URL(str(other))
+            return URL(other._val._replace(), encoded=True)
+
+        def raw(uo: URL):
+            attrs = (
+                "raw_authority",
+                "raw_path",
+                "raw_query_string",
+                "raw_fragment",
+            )
+            return {k[4:]: getattr(uo, k) or "" for k in attrs}
```

Comment:
```diff
diff --git a/yarl/_url.py b/yarl/_url.py
index fb75339..7e4cfeb 100644
--- a/yarl/_url.py
+++ b/yarl/_url.py
@@ -1107,6 +1107,14 @@ class URL:
             name = name[: -len(old_suffix)] + suffix
         return self.with_name(name)
 
+    def _raw_dict(self):
+        return {
+            "authority": self.raw_authority,
+            "path": self.raw_path,
+            "query_string": self.raw_query_string,
+            "fragment": self.raw_fragment
+        }
+
     def join(self, url):
         """Join URLs
 
@@ -1128,21 +1136,12 @@ class URL:
         if scheme != self.scheme or scheme not in uses_relative:
             return URL(other._val._replace(), encoded=True)
 
-        def raw(uo: URL):
-            attrs = (
-                "raw_authority",
-                "raw_path",
-                "raw_query_string",
-                "raw_fragment",
-            )
-            return {k[4:]: getattr(uo, k) or "" for k in attrs}
-
-        parts = raw(self)
+        parts = self._raw_dict()
         parts["scheme"] = scheme
 
         # scheme is in uses_authority as uses_authority is a superset of uses_relative
         if scheme in uses_authority and other.authority:
-            parts.update(raw(other))
+            parts.update(other._raw_dict())
             return URL.build(**parts, encoded=True)
 
         if other.path or other.fragment:
```

It be a lot more readable (and probably a bit faster) if you wrote it something like this.

Reply:
@bdraco I'd make it a property, then.

Reply:
Plus a docstring and type annotations..

Reply:
Looks like we are missing a lot of types internally .. but we have them in the pyi.... probably nicer to clean that up so the code tools give us better visiblity.

I'll do that in another PR